### ISSUE TITLE
`change` does nothing (instead of failing) if terms involving primitive projections are not convertible

### ISSUE THREAD

**Issue by samuelgruetter** [coq/coq#14084]
```coq
Set Primitive Projections.

Record Pair {A B: Type} := mk { Fst : A; Snd : B }.
Arguments Pair : clear implicits.
Arguments mk {A B} _ _.

Goal let p := mk 4 5 in Fst p = 4.
Proof.
  intros.
  unfold Fst.
  subst p.

  (* fails with "Error: No matching clauses for match.", but should succeed: *)
  lazymatch goal with
  | [ |- Fst ?q = _ ] => idtac q
  end.

  (* succeeds with no effect, but should fail with "Error: Not convertible." *)
  lazymatch goal with
  | [ |- ?a = _ ] => change a with 42
  end.

  (* succeeds with no effect, but should turn goal into `4 = 4` *)
  lazymatch goal with
  | [ |- ?a = _ ] => change a with 4
  end.
```

Note: If I comment out `unfold Fst.`, it behaves as desired.

Coq version: Both 8.13.0 and master are affected.

@ppedrot could this be a bug in what [you called](https://github.com/coq/coq/issues/11641#issuecomment-590117395) the "infamous but well-named `pattern_of_constr` OCaml function? If so, could Ltac2 get a new `better_pattern_of_constr` that "just works"? I actually encountered this bug while working in Ltac2, and the above code also fails when translated to Ltac2, but maybe Ltac2 could get a new better `change` that's not fully compatible with Ltac1 `change`, but more correct?

**Comment by ppedrot** [comment]
> could Ltac2 get a new better_pattern_of_constr

Don't count too much on that. Pattern selection is a notably complex problem without a silver bullet solution. There are different needs that can't really be reconciled with a single algorithm, e.g. interactive use vs. tactic programming. Many people seem to consider that the ssreflect algorithm is a good replacement candidate, but even if we were to use it there are quite a few design and engineering issues on the way.

In this particular case it looks like we're simply missing the perennial hardwired quotient between folded and unfolded projections in the matching algorithm. It's probably easier (somewhat...) to solve.

**Comment by JasonGross** [comment]
> Pattern selection is a notably complex problem without a silver bullet solution. There are different needs that can't really be reconciled with a single algorithm, e.g. interactive use vs. tactic programming.

This should be documented somewhere (perhaps the wiki, or a new issue?). As far as I'm aware, there are the following issues:
- which evars are holes (can be unified with anything), which are proper evars (to be unified only with themselves)? (Perhaps also a third category: evars which can be instantiated by pattern matching)
- when matching evars, do the evar instances have to match? (Syntactically, or modulo conversion, or modulo unification where we can instantiate evars in the substitutions?)
- which universe variables are holes (unconstrained), which are to be treated for equality, and which ones impose non-equality constraints?
- should non-linear matching be done up to syntactic equality or up to conversion?
- should non-linear pattern matching be done ignoring universes, up to preexisting universe equality, or should it impose new constraints?  Or should it only match when new constraints would be compatible with the existing ones, without actually imposing those constraints?

Am I missing any?

Note that many of these go away when we forbid non-linear pattern matching.

**Comment by ppedrot** [comment]
Fixed by #15559.
